### PR TITLE
Make it possible to build with `make --warn-undefined-variables`

### PR DESCRIPTION
**Pull request by thockin**
/kind cleanup

When you run our build with undefined variable warnings in make, it flags a bunch of errors.  In digging some of these are real bugs.  After this set of commits, it is possible to rebuild k/k with `--warn-undefined-variables` and get no errors.  I specifically said "rebuild".  The initial build still flags errors since auto-generated variables have not yet been defined.  I'll pursue that separately.

I suggest reading this commit-by-commit.  The biggest change is renaming code-generator's "_examples" dir, and that can be discussed if need be.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/code-generator/examples/_examples/HyphenGroup/informers/externalversions/internalinterfaces/BUILD [hunk 1]
```diff
@@ -0,0 +1,15 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
```
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/examples/_examples/HyphenGroup/informers/externalversions/internalinterfaces/BUILD:1]: Are files in `.../examples/_examples/...` expected?

**Comment by lavalamp** [comment]
I think the rename somehow got a duplicate nested _examples, can you check that? the other commits look good.

**Comment by thockin** [comment]
Woops, bad rename indeed.  Sorry

**Comment by liggitt** [comment]
The dependency failure is legitimate, dropping the underscore makes the examples dir visible to go and sets up a module cycle

**Comment by thockin** [comment]
I'll look at all the fails, but I am missing the import cycle - where is that?

If we do not rename this, the alternatives I see are:

1) special case `staging/src/k8s.io/code-generator/_examples`
2) run codegen on any `_name` directory, which is just _examples in practice, but future

**Comment by liggitt** [comment]
> I'll look at all the fails, but I am missing the import cycle - where is that?

Oh, hrmm... I thought it was a cycle at the module level, but I guess it's not.

It does make k8s.io/code-generator depend on `k8s.io/{api,apimachinery,client-go}`, which will ripple to anyone consuming `k8s.io/code-generator`. Running update-vendor.sh makes this explicit:

```diff
diff --git a/staging/src/k8s.io/code-generator/go.mod b/staging/src/k8s.io/code-generator/go.mod
index 52aa67052e8..c0ac68c27c9 100644
--- a/staging/src/k8s.io/code-generator/go.mod
+++ b/staging/src/k8s.io/code-generator/go.mod
@@ -6,20 +6,23 @@ go 1.15
 
 require (
 	github.com/emicklei/go-restful v2.9.5+incompatible // indirect
+	github.com/go-openapi/spec v0.19.3
 	github.com/gogo/protobuf v1.3.1
-	github.com/google/go-cmp v0.5.2 // indirect
-	github.com/json-iterator/go v1.1.10 // indirect
 	github.com/mailru/easyjson v0.7.0 // indirect
 	github.com/spf13/pflag v1.0.5
-	github.com/stretchr/testify v1.6.1 // indirect
 	golang.org/x/mod v0.3.0 // indirect
-	golang.org/x/net v0.0.0-20201110031124-69a78807bb2b // indirect
-	golang.org/x/text v0.3.4 // indirect
 	golang.org/x/tools v0.0.0-20200616133436-c1934b75d054 // indirect
-	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	k8s.io/api v0.0.0
+	k8s.io/apimachinery v0.0.0
+	k8s.io/client-go v0.0.0
 	k8s.io/gengo v0.0.0-20201214224949-b6c5ce23f027
 	k8s.io/klog/v2 v2.4.0
 	k8s.io/kube-openapi v0.0.0-20201113171705-d219536bb9fd
 )
 
-replace k8s.io/code-generator => ../code-generator
+replace (
+	k8s.io/api => ../api
+	k8s.io/apimachinery => ../apimachinery
+	k8s.io/client-go => ../client-go
+	k8s.io/code-generator => ../code-generator
+)
```

Technically, k8s.io/{api,apimachinery,client-go} probably _should_ depend on k8s.io/code-generator, and their generation scripts should actually live beneath their source trees, but we sort of cheat by having the generation code live in the k8s.io/kubernetes root.

**Comment by thockin** [comment]
> It does make k8s.io/code-generator depend on k8s.io/{api,apimachinery,client-go}, which will ripple to anyone consuming k8s.io/code-generator

I mean, it already DOES depend on those, it's just "hidden" in _examples.

Do you object to updating the vendoring on this?

**Comment by liggitt** [comment]
> > It does make k8s.io/code-generator depend on k8s.io/{api,apimachinery,client-go}, which will ripple to anyone consuming k8s.io/code-generator
> 
> I mean, it already DOES depend on those, it's just "hidden" in _examples.

The examples dir depends on those... but we don't want to expose all consumers of k8s.io/code-generator to dependencies they don't need (specifically k8s.io/api and k8s.io/client-go...)

> Do you object to updating the vendoring on this?

I would probably isolate the `examples` dir with something like this:

```diff
diff --git a/staging/src/k8s.io/code-generator/examples/go.mod b/staging/src/k8s.io/code-generator/examples/go.mod
new file mode 100644
index 00000000000..5a969627e13
--- /dev/null
+++ b/staging/src/k8s.io/code-generator/examples/go.mod
@@ -0,0 +1,5 @@
+// This is a submodule to isolate k8s.io/code-generator from k8s.io/{api,apimachinery,client-go} dependencies in generated code
+
+module k8s.io/code-generator/examples
+
+go 1.15

```

**Comment by lavalamp** [comment]
+1 to not expanding the dependency graph.

**Comment by thockin** [comment]
Do we have any automation around go.mod management OTHER THAN k/k/go.mod and k/k/staging/.../go.mod ?

**Comment by thockin** [comment]
What I mean is:

```
/staging/src/k8s.io/code-generator/examples$ go mod tidy
go: finding module for package github.com/googleapis/gnostic/OpenAPIv2
go: finding module for package k8s.io/api/settings/v1alpha1
go: finding module for package k8s.io/api/auditregistration/v1alpha1
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	github.com/googleapis/gnostic/OpenAPIv2: module github.com/googleapis/gnostic@latest found (v0.5.3), but does not contain package github.com/googleapis/gnostic/OpenAPIv2
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	k8s.io/client-go/kubernetes/scheme imports
	k8s.io/api/auditregistration/v1alpha1: module k8s.io/api@latest found (v0.20.2), but does not contain package k8s.io/api/auditregistration/v1alpha1
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	k8s.io/client-go/kubernetes/scheme imports
	k8s.io/api/settings/v1alpha1: module k8s.io/api@latest found (v0.20.2), but does not contain package k8s.io/api/settings/v1alpha1
```

**Comment by thockin** [comment]
I have reached the end of my comprehension of go modules.  Manually adding require/replace does not, in fact, require or replace.

```
$ cat go.mod 
// This is a submodule to isolate k8s.io/code-generator from k8s.io/{api,apimachinery,client-go} dependencies in generated code

module k8s.io/code-generator/examples

go 1.15

require (
	github.com/go-openapi/spec v0.20.1
	github.com/googleapis/gnostic v0.4.1 // indirect
	golang.org/x/time v0.0.0-20201208040808-7e3f01d25324 // indirect
	k8s.io/api v0.20.2
	k8s.io/apimachinery v0.20.2
	k8s.io/client-go v11.0.0+incompatible
	k8s.io/klog v1.0.0 // indirect
	k8s.io/kube-openapi v0.0.0-20210113233702-8566a335510f
	k8s.io/utils v0.0.0-20210111153108-fddb29f9d009 // indirect
)

replace github.com/googleapis/gnostic => github.com/googleapis/gnostic v0.4.1
thockin@thockin-glaptop4 kubernetes makefile-unused-variables-watning /staging/src/k8s.io/code-generator/examples$ go mod tidy
go: finding module for package k8s.io/api/settings/v1alpha1
go: finding module for package k8s.io/api/auditregistration/v1alpha1
go: finding module for package github.com/googleapis/gnostic/OpenAPIv2
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	github.com/googleapis/gnostic/OpenAPIv2: module github.com/googleapis/gnostic@latest found (v0.5.3, replaced by github.com/googleapis/gnostic@v0.4.1), but does not contain package github.com/googleapis/gnostic/OpenAPIv2
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	k8s.io/client-go/kubernetes/scheme imports
	k8s.io/api/auditregistration/v1alpha1: module k8s.io/api@latest found (v0.20.2), but does not contain package k8s.io/api/auditregistration/v1alpha1
k8s.io/code-generator/examples/HyphenGroup/clientset/versioned imports
	k8s.io/client-go/discovery imports
	k8s.io/client-go/kubernetes/scheme imports
	k8s.io/api/settings/v1alpha1: module k8s.io/api@latest found (v0.20.2), but does not contain package k8s.io/api/settings/v1alpha1
```

**Comment by liggitt** [comment]
we probably want the examples go.mod file to reference the local k8s.io/{api,apimachinery,client-go} content (rather than immediately stale prior published versions), so this would work:

```diff
diff --git a/staging/src/k8s.io/code-generator/examples/go.mod b/staging/src/k8s.io/code-generator/examples/go.mod
new file mode 100644
index 00000000000..81fd49ae130
--- /dev/null
+++ b/staging/src/k8s.io/code-generator/examples/go.mod
@@ -0,0 +1,19 @@
+// This is a submodule to isolate k8s.io/code-generator from k8s.io/{api,apimachinery,client-go} dependencies in generated code
+
+module k8s.io/code-generator/examples
+
+go 1.15
+
+require (
+       github.com/go-openapi/spec v0.19.3
+       k8s.io/api v0.0.0
+       k8s.io/apimachinery v0.0.0
+       k8s.io/client-go v0.0.0
+       k8s.io/kube-openapi v0.0.0-20201113171705-d219536bb9fd
+)
+
+replace (
+       k8s.io/api => ../../api
+       k8s.io/apimachinery => ../../apimachinery
+       k8s.io/client-go => ../../client-go
+)
```

**Comment by thockin** [comment]
OK.  New go module added (thanks @liggitt) and some other debug that was valuable.  I suspect I will have to edit some import controls still, but I wanted to get a full signal on the new state.

File: hack/verify-internal-modules.sh [hunk 2]
```diff
@@ -31,14 +31,14 @@ mkdir -p "${_tmp_hack}/.."
 cp -a "${KUBE_ROOT}/hack" "${_tmp_hack}/.."
 
 pushd "${_tmp_kuberoot}" >/dev/null
-./hack/update-hack-tools.sh
+./hack/update-internal-modules.sh
 popd
 
 diff=$(diff -Naupr "${KUBE_ROOT}/hack" "${_tmp_kuberoot}/hack" || true)
```
Comment (liggitt) [inline hack/verify-internal-modules.sh:37]: this only diffs the `hack` dir, but `hack/update-internal-modules.sh` now modifies stuff under `hack` and `staging`

**Comment by liggitt** [comment]
mostly nits on the code-generator examples bit. cc @sttts for visibility

I didn't dig into the generated makefile changes

File: hack/verify-internal-modules.sh (hunk 2 above)
Comment (thockin) [reply hack/verify-internal-modules.sh:37]: doh, will fix

Comment (thockin) [reply hack/verify-internal-modules.sh:37]: Will fix by switching to use git worktree - faster AND more complete.

**Comment by thockin** [comment]
@liggitt I don't see any nits from you except the broken verify.  Did you have more and forget to send?

**Comment by liggitt** [comment]
> @liggitt I don't see any nits from you except the broken verify. Did you have more and forget to send?

I thought I did, but... nope, I guess that was it...

**Comment by thockin** [comment]
new push is up

**Comment by thockin** [comment]
go lint fails but doesn't say why:

```
installing golint
go: downloading golang.org/x/lint v0.0.0-20200302205851-738671d3881b
go: downloading golang.org/x/tools v0.0.0-20200710042808-f1c4188a97a1
+++ exit code: 1
```

**Comment by thockin** [comment]
sigh, bug in our golint script.  Will add a commit to fix that early in the stack

**Comment by thockin** [comment]
Now with more golint

**Comment by liggitt** [comment]
lgtm

**Comment by thockin** [comment]
Rebased

**Comment by sttts** [comment]
With a new dependency order (am not a fan), has the publishing order been updated as well? Otherwise, we generate wrong go.mod/sum in the bot I believe cc @nikhita

**Comment by nikhita** [comment]
> With a new dependency order (am not a fan), has the publishing order been updated as well? Otherwise, we generate wrong go.mod/sum in the bot I believe cc @nikhita

The publishing-bot only looks at the root go.mod file today and cannot handle multi-modules. Since the new dependency order is in the examples go.mod (and not the root go.mod), even after updating the publishing order, the bot will not generate the correct go.mod.

Right now, the bot will generate this for `examples/go.mod` in the published repo:

```
// This is a submodule to isolate k8s.io/code-generator from k8s.io/{api,apimachinery,client-go} dependencies in generated code

module k8s.io/code-generator/examples

go 1.15

require (
        github.com/go-openapi/spec v0.19.3
        k8s.io/api v0.0.0
        k8s.io/apimachinery v0.0.0
        k8s.io/client-go v0.0.0
        k8s.io/kube-openapi v0.0.0-20201113171705-d219536bb9fd
)

replace (
        k8s.io/api => ../../api
        k8s.io/apimachinery => ../../apimachinery
        k8s.io/client-go => ../../client-go
)
```

Are we fine with a wrong published go.mod file for the examples module since it's not meant to be imported? If yes, https://github.com/kubernetes/publishing-bot/pull/244 can help unblock publishing for now.

cc @dims

**Comment by liggitt** [comment]
> Are we fine with a wrong published go.mod file for the examples module since it's not meant to be imported?

incrementally, I think so... it was in a directory hidden to go before without any dependency information stated

**Comment by liggitt** [comment]
> With a new dependency order (am not a fan), has the publishing order been updated as well?

the goal of the submodule was to avoid touching dependency ordering for modules we intend people to use

**Pull request merged**